TiKV gRPC server coredump when TLS is enabled

## Bug Report

TiKV core dumps in BR TLS tests, see more https://github.com/pingcap/br/pull/202

### What version of TiKV are you using?

```
./tikv-server -V
TiKV
Release Version:   4.1.0-alpha
Git Commit Hash:   e364c57f69151c7fb4b00de591217f9edfedd61b
Git Commit Branch: master
UTC Build Time:    2020-03-21 07:37:31
Rust Version:      rustc 1.42.0-nightly (0de96d37f 2019-12-19)
Enable Features:   jemalloc portable sse protobuf-codec
Profile:           dist_release
```

### Steps to reproduce

Run BR integration tests

### What did you expect?

TiKV should start normally or report an error.

### What did happened?

Core dumped.

[TiDB Binlog](https://github.com/pingcap/tidb-binlog) meets this too


reproduction steps:
Int the tidb-binlog repo:
1. put  tidb/tikv/pd binary in bin/
2. run `make`
3. run `sh tests/run.sh --debug`


if normally you can see:
```
Verifying drainer is started...
{"PumpPos":{"pump:8250":415530729860235266},"Synced":false,"LastTS":415530729860235266,"TsMap":""}
You may now debug from another terminal. Press [ENTER] to continue.
```

but the tikv that enable TLS Core dumped

you can check the log in `/tmp/tidb_binlog_test/`

the upstream tikv's log end at:
```
[2020/03/25 15:49:52.692 +08:00] [INFO] [<unknown>] ["New connected subchannel at 0x7fd05b879020 for subchannel 0x7fd00ca67380"]
[2020/03/25 15:49:52.693 +08:00] [INFO] [util.rs:427] ["connecting to PD endpoint"] [endpoints=https://127.0.0.1:2379]
[2020/03/25 15:49:52.696 +08:00] [INFO] [<unknown>] ["New connected subchannel at 0x7fd05b878240 for subchannel 0x7fd00ca67380"]
[2020/03/25 15:49:52.697 +08:00] [INFO] [util.rs:486] ["connected to PD leader"] [endpoints=https://127.0.0.1:2379]
[2020/03/25 15:49:52.697 +08:00] [INFO] [util.rs:204] ["heartbeat sender and receiver are stale, refreshing ..."]
[2020/03/25 15:49:52.697 +08:00] [WARN] [util.rs:223] ["updating PD client done"] [spend=8.451843ms]
[2020/03/25 15:49:57.243 +08:00] [ERROR] [<unknown>] ["assertion failed: prior > 0"]
[2020/03/25 15:49:57.243 +08:00] [ERROR] [<unknown>] ["assertion failed: prior > 0"]
```

--- a/components/tikv_util/src/security.rs
+++ b/components/tikv_util/src/security.rs
@@ -1,9 +1,10 @@
 // Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
 
 use std::error::Error;
-use std::fs::File;
+use std::fs::{self, File};
 use std::io::Read;
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
+use std::time::SystemTime;
 
 use grpcio::{
     CertificateRequestType, Channel, ChannelBuilder, ChannelCredentialsBuilder, ServerBuilder,
@@ -94,6 +95,19 @@ impl SecurityConfig {
         }
         Ok((ca, cert, key))
     }
+
+    /// Determine if the cert file has been modified.
+    /// If modified, update the timestamp of this modification.
+    fn is_modified(&self, last: &mut Option<SystemTime>) -> Result<bool, Box<dyn Error>> {
+        let this = fs::metadata(&self.cert_path)?.modified()?;
+        if let Some(last) = last {
+            if *last == this {
+                return Ok(false);
+            }
+        }
+        *last = Some(this);
+        Ok(true)
+    }
 }
 
 #[derive(Default)]
@@ -134,6 +148,7 @@ impl SecurityManager {
         } else {
             let fetcher = Box::new(Fetcher {
                 cfg: self.cfg.clone(),
+                last_modified_time: Arc::new(Mutex::new(None)),
             });
             sb.bind_with_fetcher(
                 addr,
@@ -151,18 +166,31 @@ impl SecurityManager {
 
 struct Fetcher {
     cfg: Arc<SecurityConfig>,
+    last_modified_time: Arc<Mutex<Option<SystemTime>>>,
 }
 
 impl ServerCredentialsFetcher for Fetcher {
+    // Retrieves updated credentials. When returning `None` or
+    // error, gRPC will continue to use the previous certificates
+    // returned by the method.
     fn fetch(&self) -> Result<Option<ServerCredentialsBuilder>, Box<dyn Error>> {
-        let (ca, cert, key) = self.cfg.load_certs()?;
-        let new_cred = ServerCredentialsBuilder::new()
-            .add_cert(cert, key)
-            .root_cert(
-                ca,
-                CertificateRequestType::RequestAndRequireClientCertificateAndVerify,
-            );
-        Ok(Some(new_cred))
+        if let Ok(mut last) = self.last_modified_time.try_lock() {
+            // Reload only when cert is modified.
+            if self.cfg.is_modified(&mut last)? {
+                let (ca, cert, key) = self.cfg.load_certs()?;
+                let new_cred = ServerCredentialsBuilder::new()
+                    .add_cert(cert, key)
+                    .root_cert(
+                        ca,
+                        CertificateRequestType::RequestAndRequireClientCertificateAndVerify,
+                    );
+                Ok(Some(new_cred))
+            } else {
+                Ok(None)
+            }
+        } else {
+            Ok(None)
+        }
     }
 }
 
@@ -171,7 +199,7 @@ mod tests {
     use super::*;
 
     use std::fs;
-
+    use std::io::Write;
     use tempfile::Builder;
 
     #[test]
@@ -225,5 +253,21 @@ mod tests {
         assert_eq!(ca, vec![0]);
         assert_eq!(cert, vec![1]);
         assert_eq!(key, vec![2]);
+    }
+
+    #[test]
+    fn test_modify_file() {
+        let mut file = Builder::new().prefix("test_modify").tempfile().unwrap();
+        let mut cfg = SecurityConfig::default();
+        cfg.cert_path = file.path().to_str().unwrap().to_owned();
+
+        let mut last = None;
+        assert!(cfg.is_modified(&mut last).unwrap());
+        assert!(!cfg.is_modified(&mut last).unwrap());
+
+        std::thread::sleep(std::time::Duration::from_millis(10));
+        writeln!(file, "something").unwrap();
+        assert!(cfg.is_modified(&mut last).unwrap());
+        assert!(!cfg.is_modified(&mut last).unwrap());
     }
 }

